### ISSUE TITLE
Multiple base_dir path definitions

### ISSUE THREAD

**Issue by maikelvdh** [ccache/ccache#597]
In order to support package managers better, which write their retrieved data typically in unique paths, it would be nice to be able to declare multiple `base_dir` paths to translate absolute paths in relative ones.

For example a package manager would have the following paths in its cache:
```
~/data/libA/1.0.0/acme/stable/package/63da998e3642b50bee33f4449826b2d623661505/include/libA/myheaderA.h
~/data/libB/2.0.0/acme/stable/package/63da998e3642b50bee33f4449826b2d623661505/include/libB/myheaderB.h
```
You would like to eliminate in such cases the parts until the hash value directory as the part before is variable and changing when version does change. The changes in updated version is something ccache will already help with

**Comment by jrosdahl** [comment]
Supporting multiple base directories is something that has been mentioned several times through the years but it has only reached my personal todo list so far. It's good to have the idea tracked in a GitHub issue.

However, in your example it would be enough to use `$HOME/data` as the base directory, right?

**Comment by maikelvdh** [comment]
If you would use `$HOME/data` in my example then you would have still no cache hits in case the version of libA or libB would change. While in theory the headers between for example libA 1.0.0 and 1.1.0 could be equal. If this version would not be part of the cached directory structure then you would have cache hits.

**Comment by jrosdahl** [comment]
It sounds like you haven't fully grasped what `base_dir` does. `base_dir` specifies a prefix. If an absolute path matches the prefix it is rewritten as relative to _the current working directory_, not to `base_dir`.

Say that you first compile like this:

```console
cd $HOME/data/libA/1.0.0/acme/stable/package/63da998e
gcc -I$HOME/data/libA/1.0.0/acme/stable/package/63da998e/include -c $HOME/data/libA/1.0.0/acme/stable/package/63da998e/foo.c
```

If you later do

```console
cd $HOME/data/libA/1.0.1/acme/stable/package/2a15201d
gcc -I$HOME/data/libA/1.0.1/acme/stable/package/2a15201d/include -c $HOME/data/libA/1.0.1/acme/stable/package/2a15201d/foo.c
```

(and the source code is identical) you won't get a cache hit because of the absolute paths. But if you would build with e.g. `CCACHE_BASEDIR=$HOME` you would get a cache hit since ccache then would rewrite the paths like this:

```console
cd $HOME/data/libA/1.0.1/acme/stable/package/2a15201d
gcc -Iinclude -c foo.c
```

Does that make things clearer? If not, please describe why you think you wouldn't get cache hits.

**Comment by maikelvdh** [comment]
Then indeed I misunderstood the usage of `base_dir`. The scenario I would like to achieve is for example:
```
cd $HOME/myproject
gcc -I$HOME/data/libA/1.0.0/acme/stable/package/2a15201d/include foo.c
```
And then later:
```
cd $HOME/myproject
gcc -I$HOME/data/libA/1.0.1/acme/stable/package/2a15201d/include foo.c
```

**Comment by jrosdahl** [comment]
OK. `base_dir` won't help you with that other scenario. You should be able to get preprocessed hits regardless of `base_dir`, though, since `-I` values are not part of the preprocessed mode hash.

**Comment by maikelvdh** [comment]
@jrosdahl is it correct that `-isystem` isn't taken into account with preprocessed mode? Also it seems that it does only work in case you have `-I<path>` and not if you have `-I <path>`. Tested it with ccache 3.7.3

**Comment by jrosdahl** [comment]
OK, right. The hashed `inc_path` value comes from `make_relative_path` which uses `base_dir`, but that won't help in the discussed scenario.

> is it just a matter of missing functionality to tune this hashing of include paths?

If it can be done in a safe way, sure. It's unfortunately tricky to foresee the consequences of these things as can be witnessed by #88/#134.

> For example passing -P to the preprocessor invocation to disabled line marker output.

`-P` was discussed in this mail thread: https://www.mail-archive.com/ccache@lists.samba.org/msg01325.html

> Similarly it might be worth to also skip for example comments in the preprocessor output to maximise the hits further, since a change of comment shouldn't lead to a cache miss I would say.

That's a reasonable desire, but compilers can behave differently depending on comments. One example is [fallthrough comments](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wimplicit-fallthrough_003d). Making such behavior opt-in would be fine, though.

**Comment by MaikelvandenHurk-tomtom** [comment]
Thanks for the detailed answer! Isn't the scenario of #134 something which gets addressed by: #592? Or do I misunderstand that?

**Comment by maikelvdh** [comment]
Thanks for the detailed answer! Isn't the scenario of #134 something which gets addressed by: #592? Or do I misunderstand that?

**Comment by jrosdahl** [comment]
> Thanks for the detailed answer! Isn't the scenario of #134 something which gets addressed by: #592? Or do I misunderstand that?

Yes, maybe. Unless it introduces a bug for the `-Wp,-MD` and `-Wp,-MMD` options which are not addressed by #592, etc. My main point is that it's easy to introduce bugs when altering which parts of the preprocessor output to hash so one has to be extra careful to think through the consequences.
